sentry-android-core/src/main/java/io/sentry/android/core/SpanFrameMetricsCollector.java

```diff
@@ -1,43 +1,82 @@
 import io.sentry.protocol.MeasurementValue;
+import io.sentry.util.Objects;
 import java.util.HashMap;
+import java.util.Iterator;
 import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.concurrent.ConcurrentSkipListSet;
 import java.util.concurrent.TimeUnit;
 import org.jetbrains.annotations.ApiStatus;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.TestOnly;
 
 @ApiStatus.Internal
 public class SpanFrameMetricsCollector
     implements IPerformanceContinuousCollector,
         SentryFrameMetricsCollector.FrameMetricsCollectorListener {
 
-  private @NotNull final Object lock = new Object();
-  private @Nullable final SentryFrameMetricsCollector frameMetricsCollector;
-  private @Nullable volatile String listenerId;
-  private @NotNull final Map<SpanId, SentryFrameMetrics> metricsAtSpanStart;
+  // 30s span duration at 120fps = 3600 frames
+  private static final int MAX_FRAMES_COUNT = 3600;
+  private static final long ONE_SECOND_NANOS = TimeUnit.SECONDS.toNanos(1);
 
-  private @NotNull final SentryFrameMetrics currentFrameMetrics;
+  private final @NotNull SentryAndroidOptions options;
+  private final @NotNull FrameTimeProvider frameTimeProvider;
   private final boolean enabled;
 
-  private float lastRefreshRate = 60.0f;
+  private final @NotNull Object lock = new Object();
+  private final @NotNull SentryFrameMetricsCollector frameMetricsCollector;
+
+  private volatile @Nullable String listenerId;
+
+  // a map of <span-id, span-start-time-nanos>
+  // as spans do not contain a nano start time, it's being tracked here
+  private final @NotNull Map<String, NanoTimeStamp> spanStarts;
+
+  // all running spans, sorted by span start nano time
+  private final @NotNull SortedSet<SpanStart> runningSpans;
+
+  // all collected frames, sorted by frame end time
+  // this is a concurrent set, as the frames are added on the main thread,
+  // but span starts/finish may happen on any thread
+  // the list only holds Frames, but in order to query for a specific span NanoTimeStamp is used
+  private final @NotNull ConcurrentSkipListSet<NanoTimeStamp> frames;
+
+  // assume 60fps until we get a value reported by the system
+  private long lastKnownFrameDurationNanos = 16_666_666L;
 
   public SpanFrameMetricsCollector(final @NotNull SentryAndroidOptions options) {
-    frameMetricsCollector = options.getFrameMetricsCollector();
+    //noinspection Convert2MethodRef
+    this(options, () -> System.nanoTime());
+  }
+
+  @TestOnly
+  public SpanFrameMetricsCollector(
+      final @NotNull SentryAndroidOptions options,
+      final @NotNull FrameTimeProvider frameTimeProvider) {
+    this.options = options;
+    this.frameTimeProvider = frameTimeProvider;
+
+    frameMetricsCollector =
+        Objects.requireNonNull(
+            options.getFrameMetricsCollector(), "FrameMetricsCollector is required");
```

Comment:
since `options.getFrameMetricsCollector()` is nullable, what about adding a `SentryFrameMetricsCollector` in the params of the constructor?
We have access to it anyway in the `AndroidOptionsInitializer`